### PR TITLE
File Uploader

### LINKED ISSUE #394: New component: File Uploader
**Purpose:** Create a component that allows users to upload files.

**Design:** https://cengage.invisionapp.com/share/BH10W909XG5D

### Acceptance Criteria
1. Simple file upload option
-- The component can have a title
-- The component can have a description below the title
-- There is a button labeled "Browse files"
-- Pressing the button brings up your computer's version of Finder or Explorer to find a file.
-- Selecting a file goes back to the uploader, and the uploading item is displayed below the button
-- You can see the following on the uploading item
--- Identifier icon (see inVision link for possible icons)
--- File name with extension (.png, .pdf, etc.)
---- A long file name gets truncated automatically with an ellipsis ( ... ) if it runs into the items on the right.
--- loading progress percentage
--- loading progress ring
-- Verify when the upload is complete, the loading ring transitions to a checkmark icon
-- Verify when the checkmark icon has displayed for about 1.5 seconds, it turns to a delete button with the trash icon.
-- Verify if there is an error uploading the file, an appropriate error message is displayed.  See inVision link for baked in error messages. 
2. Drag and drop option
-- The component can have a title
-- The component can have a description below the title
-- The drop zone is identified by the container with the dashed border
-- Inside the drop zone there are instructions saying "Drag and drop or browse files to upload"
-- Pressing the Browse Files button triggers the same behavior as detailed above in the first section.
-- You can drag one or more files from anywhere on your computer and drop them on the drop zone.
-- The uploading item is displayed below the drop zone
-- You can see the following on the uploading item
--- Identifier icon (see inVision link for possible icons)
--- File name with extension (.png, .pdf, etc.)
---- A long file name gets truncated automatically with an ellipsis ( ... ) if it runs into the items on the right.
--- loading progress percentage
--- loading progress ring
-- Verify when the upload is complete, the loading ring transitions to a checkmark icon
-- Verify when the checkmark icon has displayed for about 1.5 seconds, it turns to a delete button with the trash icon.
-- Verify if there is an error uploading the file, an appropriate error message is displayed.  See inVision link for baked in error messages. 
3. The uploader has an inverse option where the styling is changed so you can use it on dark backgrounds.  See design in inVision
4. Setting restrictions
-- A developer may set restrictions on what can be uploaded
--- Limit the uploads to specific file types
--- Limit the size of each uploaded item
--- Limit the total number of items that may be uploaded

### PR DESCRIPTION
**Pull request by aaronmaturen**

### CODE REVIEW

**Review by orion-cengage — APPROVED** [review]
The file uploader is looking good to me and ready to be reviewed by Perkins after we publish it.

File: packages/file-uploader/src/components/fileuploader/FileIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  FilePdfIcon,
+  FilePowerpointIcon,
+  FileWordIcon,
+  FileZipIcon,
+  IconProps,
+} from 'react-magma-icons';
+
+import { FilePreview } from './FilePreview';
+
+export interface FileIconProps extends IconProps {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/FileIcon.tsx:17]: Do these props need the typedoc information?

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+  styled,
+  useIsInverse,
+} from 'react-magma-dom';
+
+import { FileIcon } from './FileIcon';
+import { FilePreview } from './FilePreview';
+import { formatFileSize } from './utils';
+import React from 'react';
+
+export interface PreviewProps extends Omit<FlexProps, 'behavior'> {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:34]: typedoc

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      thumbnails,
+      ...rest
+    } = props;
+
+    const theme: ThemeInterface = useContext(ThemeContext);
+    const i18n: I18nInterface = React.useContext(I18nContext);
+    const isInverse = useIsInverse(isInverseProp);
+    const [actions, setActions] = useState(<CloseIcon />);
+
+    const handleRemoveFile = (file: FilePreview) => {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:180]: small thing, but in all of our other components we use the `function {x}` format

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+      if (onRemoveFile && typeof onRemoveFile === 'function') {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:181]: in all of our other components we use `{x} && typeof {x} === 'function' && {x}({y})`

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    };
+
+    const handleDeleteFile = (file: FilePreview) => {
+      if (onDeleteFile && typeof onDeleteFile === 'function') {
+        onDeleteFile(file);
+      }
+    };
+
+    const FinishedActions = ({ status = 'ready' }: { status?: string }) => {
+      const [done, setDone] = useState(false);
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:193]:
```suggestion
      const [done, setDone] = useState<boolean>(false);
```

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+      if (status === 'error' || status === 'ready') {
+        return (
+          <StatusIcons>
+            <IconButton
+              onClick={() => {
+                handleRemoveFile(file);
+              }}
+              variant={ButtonVariant.link}
+              color={ButtonColor.secondary}
+              aria-label="Remove File"
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:216]: hardcoded. needs i18n

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        return () => {
+          mounted = false;
+        };
+      }, [status]);
+
+      if (status === 'error' || status === 'ready') {
+        return (
+          <StatusIcons>
+            <IconButton
+              onClick={() => {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:211]: `file` is in `props`. No need for the `() => remove(file)`

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+      }
+
+      return (
+        <StatusIcons>
+          <Transition isOpen={!done} unmountOnExit fade>
+            <CheckCircleIcon
+              color={
+                isInverse ? theme.colors.successInverse : theme.colors.success
+              }
+              style={{ marginTop: '4px' }}
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:242]: hardcoded

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+            />
+          </Transition>
+          <Transition isOpen={done} unmountOnExit fade>
+            <IconButton
+              onClick={() => {
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:247]: no need for `() => delete(file)`

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+                handleDeleteFile(file);
+              }}
+              variant={ButtonVariant.link}
+              color={ButtonColor.secondary}
+              aria-label="Delete File"
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:252]: hardcoded. i18n

File: packages/file-uploader/src/components/fileuploader/Preview.tsx [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+                  color={
+                    isInverse ? theme.colors.dangerInverse : theme.colors.danger
+                  }
+                  size={24}
+                />
+              ) : file.preview &&
+                thumbnails &&
+                file.type &&
+                file.type.startsWith('image') ? (
+                <Thumb role="img" id="THISISTHETHUMBNAIL" file={file} />
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/Preview.tsx:296]: why the id?

File: packages/file-uploader/src/components/fileuploader/utils.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+export const formatFileSize = (bytes: number | undefined, decimalPoint?:number) => {
+  if(bytes === undefined) return;
+  if(bytes == 0) return '0 Bytes';
```
Comment (rcgillet) [inline packages/file-uploader/src/components/fileuploader/utils.ts:3]: not 100%, but does `Bytes` need to be added to the i18n stuff?

**Review by rcgillet — CHANGES_REQUESTED** [review]

File: packages/file-uploader/src/components/fileuploader/FilePreview.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import { FileWithPath } from 'react-dropzone';
+
+export interface FileError {
```
[the remark this answers is not in the document]
Comment (aaronmaturen) [reply packages/file-uploader/src/components/fileuploader/FilePreview.ts:3]: just internal again

**Review by orion-cengage — CHANGES_REQUESTED** [review]
Perkins Testing:
1. The text "Upload files" is implemented as a label element for an input that is hidden via display:none. When this text is clicked it opens the file browser, which is unexpected because the text does not look like an actionable control. Consider removing the label element from this text and use a span instead.
2. The div that contains the drop zone for the file uploader has a tabindex="0" attribute making it focusable with the keyboard. When this region gains focus it does not show a visible focus indicator. This could be confusing for keyboard users. Since there is a "Browse Files" button that gains keyboard focus and opens the file chooser, the div does not also need to gain focus. Therefore consider removing the tabindex attribute.
image

3. After uploading a file, screen readers can access content with role="progressbar" that is not visible. This progressbar has a value of "zero percent" which makes it seem as though the file has yet to be uploaded. If this element is not meant to be visible it should not be detected by screen readers. Otherwise if it is meant to communicate the progress of the uploaded file it should show some visual indication in the UI as well as in the code.
image

**Comment by aaronmaturen** [comment]
fixed the issues from perkins

**Review by orion-cengage — APPROVED** [review]

**Pull request merged**